### ISSUE TITLE
New check: variable defined away from use

### ISSUE THREAD

**Issue by nickdrozd** [astral-sh/ruff#15477]
Sometimes a variable is defined away from its first use. If an early return comes between the definition and the use, then the definition is wasteful.

For example, this:

```python
def f():
    val = get_val()

    if condition():
        return

    do_something(val)
```

Should be changed to this:

```python
def f():
    if condition():
        return

    val = get_val()

    do_something(val)
```

Also applies to continues, etc:

```python
def f():
   for item in get_list():
       val = get_val(item)

       if condition(item):
           continue

       do_something(val)
```

Might be autofixable, but definitely unsafe.

**Comment by dylwil3** [comment]
This is an interesting idea for a rule! it will probably require better control flow analysis than is currently available, so it may have to wait a bit.

**Comment by MichaReiser** [comment]
The other challenge that I see is that it has to know if the called method has any side effects or moving it isn't safe.

**Comment by nickdrozd** [comment]
Could put check behind new flag `--experimental` or `--aggressive`, something like that. Some kind of extra opt-in.

I would be interested to see the results of this run against primer or whatever. I think this problem might turn out to be pretty common. False positives due to side effects would also interesting to see.

**Comment by nickdrozd** [comment]
Somewhat related to https://github.com/astral-sh/ruff/issues/10108.
